Update EIP-7620: Update spec (#8361)
ethereum/EIPs: EIPS/eip-5792.md

---
eip: 5792
title: Wallet Call API
description: Adds JSON-RPC methods for sending multiple calls from the user's wallet, and checking their status
author: Moody Salem (@moodysalem), Lukas Rosario (@lukasrosario), Wilson Cusack (@wilsoncusack), Dror Tirosh (@drortirosh)
discussions-to: https://ethereum-magicians.org/t/eip-5792-wallet-abstract-transaction-send-api/11374
status: Draft
type: Standards Track
category: Interface
created: 2022-10-17
---

## Abstract

> Defines new JSON-RPC methods for apps to send batches of calls from a user's wallet, as well as check on the status of those calls. The new methods have a simple core while enabling additional transaction features (e.g. sponsored transactions with a paymaster) via the introduction of capabilities. To support this, there is also a new method for querying for the capabilities a given wallet supports. The new methods are more abstract in regard to the underlying transactions than the existing transaction sending APIs to allow for differences between wallet implementations, e.g. smart accounts utilizing [ERC-4337](./eip-4337.md) or EOA wallets that support batched transactions via [EIP-3074](./eip-3074.md). Apps may use this more abstract interface to support different kinds of wallets without additional effort, as well as provide a better UX for sending batches of function calls (e.g. [ERC-20](./eip-20.md) `approve` followed by a contract call).

Reviewer: Not perfect, but I think this would improve 

```suggestion
Defines new JSON-RPC methods which enable apps to ask a wallet to process a batch of onchain write calls and to check on the status of those calls. Calls can be specified to make use of various wallet-defined capabilities, and there is also a new RPC to enable apps to discover which capabilities a wallet supports. 
```

## Motivation

The current methods used to send transactions from the user wallet and check their status are `eth_sendTransaction` and `eth_getTransactionReceipt`.

The current methods used to send transactions from the user wallet and check their status do not meet modern developer demands and cannot accomodate new transaction formats. Even the name–`eth_sendTransaction`–is an artifact of a time when nodes served as wallets. 

Today, developers want to send batch transactions, which many smart accounts can support atomically. Developers also want to make use of features afforded by new transaction formats, like paymasters in [ERC-4337](./eip-4337.md) transactions. `eth_sendTransaction` offers no way to do these things. 

> In updating to a new set of `wallet_` RPCs, our main goals are to enforce a clean separation of wallet and app concerns, and create a clear way for more features to be added over time with minimal coordination.

Reviewer: ```suggestion
In updating to a new set of `wallet_` RPCs, our main goals are to enforce a clean separation of wallet and app concerns, enable developers to make use of things like paymasters and batch transactions, and to create a clear way for more features to be added over time with minimal coordination.
```

description: Adds JSON-RPC methods for sending multiple calls from the user's wallet, and checking their status
author: Moody Salem (@moodysalem), Lukas Rosario (@lukasrosario), Wilson Cusack (@wilsoncusack), Dror Tirosh (@drortirosh), Jake Moxey (@jxom)
discussions-to: https://ethereum-magicians.org/t/eip-5792-wallet-abstract-transaction-send-api/11374
status: Draft
type: Standards Track
category: Interface
created: 2022-10-17
---

## Abstract

Defines new JSON-RPC methods which enable apps to ask a wallet to process a batch of onchain write calls and to check on the status of those calls. Calls can be specified to make use of various wallet-defined capabilities, and there is also a new RPC to enable apps to discover which capabilities a wallet supports. 

## Motivation

The current methods used to send transactions from the user wallet and check their status are `eth_sendTransaction` and `eth_getTransactionReceipt`.

The current methods used to send transactions from the user wallet and check their status do not meet modern developer demands and cannot accomodate new transaction formats. Even the name–`eth_sendTransaction`–is an artifact of a time when nodes served as wallets. 

> Today, developers want to send batch transactions, which many smart accounts can support atomically. Developers also want to make use of features afforded by new transaction formats, like paymasters in [ERC-4337](./eip-4337.md) transactions. `eth_sendTransaction` offers no way to do these things. 

Reviewer 2: It'd be a good though-experiment if this can also support EIP-3074

Reviewer: Should definitely be able to!

Reviewer 3: gm gm sers! Batched `AUTHCALL`s are definitely doable here. Happy to collaborate where I can as a rep for 3074.

In updating to a new set of `wallet_` RPCs, our main goals are to enforce a clean separation of wallet and app concerns, enable developers to make use of things like paymasters and batch transactions, and to create a clear way for more features to be added over time with minimal coordination.

## Specification

Four new JSON-RPC methods are added. Apps may begin using these methods immediately, falling back to `eth_sendTransaction` and `eth_getTransactionReceipt` when they are not available.

We define one capability related to call execution atomicity.

> ### `wallet_getCapabilities`

Reviewer 2: I suggest to remove this method and instead to reference CAIPs (which you can now reference from EIPs/ERCs the way I did when I [updated this same EIP a few months back](https://github.com/arein/dapp-smart-account/pull/1/files#diff-2ce6a27d54a5083f40751babba9d77c5529f0515cd5f8dc31a379406d52ff6cdR38).

This solution I argue is better as it doesn't ever require a roundtrip to discover the capabilities.

Reviewer 4: What do you mean by “context switch”? 

`eth_getCapabilities` should mostly be a synchronous call just like `eth_accounts`. CAIP-25 is an orthogonal abstraction to this, that said,  consumers + wallets using CAIP-25 could include `eth_getCapabiltities` in their required/optional methods.

Reviewer 4: What do you mean by “context switch”?

`eth_getCapabilities` should mostly be a synchronous call just like `eth_accounts`. CAIP-25 is an orthogonal abstraction to this, that said, consumers + wallets using CAIP-25 could include `eth_getCapabilities` in their required/optional methods.

Reviewer 5: Anything capabilities or permissions related should not be on this EIP. The Permission Standard is still in discussion and this should wait for that to move forward, since this seems like it would be dependent on the outcome of the Permission Standard. This was originally an EIP for batched function calls, imho it should stay that way, unless you can find a way to remove the capabilities, for the time being. I think it would be best to wait for Permissions Standard to move forward first.

Reviewer: @McOso what permissions standard are you referring to?

If the one Derek and others are working on, we became aware of that after we had already published our plans here. I think we have different goals and designs, and I don't think it's fair to ask us to wait on something someone else is working on which is independent of what we're doing :)  

If we can find a common best path that's great. But doesn't seem to be the case so far.

If you are sincerely concerned on the change to this EIP, the alternative path was a new proposal with these same RPCs. We took over this EIP because of the similarity in the RPCs and Moody asked us to. Otherwise this EIP is stagnant and nothing is going to happen with it. 

We should not solve for batch transactions as a one off. We need to fix several things here.

Reviewer 4: The outcome of any "permission standard" EIP would fit into this type of method, no?

IIRC, the goal of the "permission standard" is to define a standard for a consumer to **asynchronously request permissions**,
Whereas, the goal of this method is to define a standard for a consumer to **synchronously discover capabilities**.

If there is a way to synchronously discover capabilities in the permissions standard discussions, then I am all ears and maybe these efforts should be combined!

Reviewer 2: This standard needs no mention of the WIP session key/permissions standard @McOso is referring to.

For session keys the dApp will send the UserOps/Transactions. They can still use these new RPC methods to do so.

In fact I see these two things in harmony.

The [UserOpBuilder ERC](https://zerodev.notion.site/ERC_UserOpBuilder-753e73db33bd4c8c8a9fbeeaf160a27c?pvs=4) may deserve a mention here. But since it's just going into draft this week it's probably better not to.

RPC for an application to receive information about the capabilities that a connected wallet supports (e.g. batch transactions, paymaster communication).

The key of each item per chain is the name of a capability and the value can be of any shape. Capabilities are returned per chain because wallets may support different capabilities across chains.

We expect the community to align on the definition of capabilities in separate ERCs.

#### `wallet_getCapabilities` RPC Specification

```typescript
type GetCapabilitiesParams = [];

> type GetCapabilitiesResult = Record<`0x${string}`, <Record<string, any>>; // Hex chain id

Reviewer 4: We should be avoiding the `any` type here. Ideally it should be narrowed, and if not, `unknown`.

```

##### `wallet_getCapabilities` Example Return Value

The capabilities below are for illustrative purposes.

```json
{
  "0x2105": {
    "paymasterService": {
      "supported": true
    },
    "sessionKeys": {
      "supported": true

>     }

Reviewer 2: what does this mean? Should this not refer to other EIPs? e.g. `eth_sendUserOperation` to me depicts support for Paymaster.

Session keys is the TBD standard we are discussing in Telegram.

Reviewer: I think this is just an example of how wallet capabilities could be formatted. But maybe should be a different example as we don't specify any more on this, here.

Reviewer 4: What if the wallet doesn’t support `eth_sendUserOperation`, but supports sending a paymaster via `wallet_sendTransaction`.

Detecting capability support via RPC methods seems like a leaky abstraction to me.

  },
  "0x14A34": {
    "paymasterService": {
      "supported": true

>     }

Reviewer: As in you're concerned about bytes on the wire? I think it's worth it for clarity. Also at one point I think we had a description field here @lukasrosario ?

>     }

Author: yeah i think the clarity is worth it here, savings would be pretty minimal. no description field here, dont see the value in that since this is just for wallet <> app communication. maybe misunderstanding though so lmk if you think we do need something else here

>   }

Author: this is something we went back and forth on. i think this is nice because you can save a round trip if your app is multichain. but yeah have gotten this feedback before, maybe we allow for both -- i.e. without a chain id to get all chains, and with a chain id to get capabilities for just the request chain id.

@jxom curious if you have thoughts on hex chain id as keys. viem / wagmi use numbers but i believe this could be problematic because chain id can be uint256? and ts makes a fuss about bigint as keys.

>   }

Reviewer: > maybe we allow for both -- i.e. without a chain id to get all chains, and with a chain id to get capabilities for just the request chain id.

I think I would prefer to be more opinionated. @moodysalem is your main concern just how chainID is formatted? Isn't this already fairly standardized for other requests?

The capabilities field is how an app can communicate with a wallet about capabilities a wallet supports. For example, this is where an app can specify a paymaster service URL from which an [ERC-4337](./eip-4337.md) wallet can request a `paymasterAndData` input for a user operation.

The wallet:

* MUST send these calls in the order specified in the request.
* MUST send the calls on the request chain ID.
* MUST NOT send any calls from the request if the user rejects the request
* MAY revert all calls if any call fails
* MAY send all the function calls as part of one transaction or multiple transactions, depending on wallet capability.
* SHOULD stop executing the calls if any call fails
* MAY reject the request if the request chain ID does not match the currently selected chain ID.
* MAY reject the request if the from address does not match the enabled account.
* MAY reject the request if one or more calls in the batch is expected to fail, when simulated sequentially

#### `wallet_sendCalls` RPC Specification

```typescript
type SendCallsParams = {
  version: string;

>   chainId: `0x{string}`; // Hex chain id

Reviewer 6: nit: dropped a $ here

  from: `0x${string}`;
  calls: {

>     to: `0x${string}`;

Reviewer 4: ```suggestion
    to?: `0x${string}` | undefined;
```

can deploy contract?

>     data: `0x${string}`;

Reviewer 4: ```suggestion
    data?: `0x${string}` | undefined;
```

>     value: `0x${string}`; // Hex value

Reviewer 4: ```suggestion
    value?: `0x${string}` | undefined; // Hex value
```

  }[];

>   capabilities: Record<string, any>;

Reviewer 4: ```suggestion
  capabilities?: Record<string, any> | undefined;
```

```json
"0xe670ec64341771606e55d6b4ca35a1a6b75ee3d5145a99d05921026d1527331"
```

### `wallet_getCallsStatus`

Returns the status of a call batch that was sent via `wallet_sendCalls`. The identifier of the transaction is the value returned from the `wallet_sendCalls` RPC. Note this method only returns a subset of fields that `eth_getTransactionReceipt` returns, excluding any fields that may differ across wallet implementations.

* If a wallet does not execute multiple calls atomically (i.e. in multiple transactions), the receipts in the `receipts` field MUST be in order of the calls sent.
* If a wallet executes multiple calls atomically (i.e. in a single transaction), `wallet_getCallsStatus` MUST return a single receipt, corresponding to the transaction in which the calls were included.

#### `wallet_getCallsStatus` RPC Specification

```typescript
type GetCallsParams = string

type GetCallsResult = {
  status: 'PENDING' | 'CONFIRMED'
  receipts?: {
    logs: {
      address: `0x${string}`
      data: `0x${string}`
      topics: `0x${string}`[]
    }[]

>     success: boolean

Reviewer 4: Should we avoid flags like this? Should we align with the `status` in the receipt schema in the execution specs? https://github.com/ethereum/execution-apis/blob/main/src/schemas/receipt.yaml#L110-L113

Reviewer: Aligning with status spec seems good. @lukasrosario?

Author: sgtm. will update here

We initially proposed that multiple calls must be executed atomically, but after some debate we ultimately decided this was too opinionated. Instead, we are including a specification for an `atomicBatch` capability. This allows for EOA wallets to accept multiple calls and still gives developers the option to only submit batched calls if they are executed atomically.

### Call Gas Limit

Our initial proposal included an optional `gas` field for each call in the `calls` field accepted by the `walletSendCalls` method. However, we realized this could be misleading because in the case of [ERC-4337](./eip-4337.md) wallets you cannot specify a gas limit per call, only a single gas limit for all calls in the user operation. We then proposed a single `gas` value that would apply to all of the calls. This works for [ERC-4337](./eip-4337.md) wallets, but not for EOA wallets. When we decided that EOA wallets should be able to handle multiple calls, we decided to just remove the `gas` field altogether.

## Backwards Compatibility

> Wallets that do not support the following methods should return error responses to the new JSON-RPC methods. Apps MAY attempt to send the same batch of calls via `eth_sendTransaction` when they receive a not implemented call, or otherwise indicate to the user that their wallet is not supported.

Reviewer 2: In your proposal an app needs to call `wallet_getCapabilities` first. And handling errors if it's not supported. That's an extra roundtrip and odd API (error handling vs. explicit).

Following my suggestion above that would be saved.

Reviewer: Can you make clear your suggestion: what call will the app make to the wallet in order to discover capabilities?

Reviewer: FWIW I am open to the idea that `eth_requestAccounts` as the implicit connection method should possibly be replaced by some like `wallet_connect` RPC that shares accounts + capabilities + other info. Don't necessarily want to block this EIP on getting to clarity there, but if we can align on something better quickly we should!

Reviewer 2: Yes I saw you commented on https://github.com/arein/dapp-smart-account/pull/1/files#diff-2ce6a27d54a5083f40751babba9d77c5529f0515cd5f8dc31a379406d52ff6cdR38 and I linked the reuest there as you asked.

Just making very clear that this has nothing to do with _WalletConnect_. _WalletConnect_ has many things, one of them is a transport layer for remote wallets.

Our SDKs use the transport layer and support CAIPs first class. But many browser extension wallets use CAIPs without using our transport layer (even though some do). Our SDKs will also support other transport layers.

Reviewer: Thanks. So is the rpc provider_authorize?
